src/main/scala/scynamo/ScynamoEncoder.scala

```diff
@@ -12,119 +12,150 @@ import shapeless.tag.@@
+    }
 
   implicit def listEncoder[A: ScynamoEncoder]: ScynamoEncoder[List[A]] =
-    value =>
-      value.zipWithIndex
-        .parTraverse { case (x, i) =>
-          ScynamoEncoder[A].encode(x).leftMap(_.map(_.push(Index(i))))
-        }
-        .map(xs => AttributeValue.builder().l(xs: _*).build())
+    seqEncoder[A].narrow
 
   implicit def vectorEncoder[A: ScynamoEncoder]: ScynamoEncoder[Vector[A]] =
-    value =>
-      value.zipWithIndex
-        .parTraverse { case (x, i) =>
-          ScynamoEncoder[A].encode(x).leftMap(_.map(_.push(Index(i))))
-        }
-        .map(xs => AttributeValue.builder().l(xs: _*).build())
+    seqEncoder[A].narrow
 
-  implicit def setEncoder[A: ScynamoEncoder]: ScynamoEncoder[Set[A]] = listEncoder[A].contramap[Set[A]](x => x.toList)
+  implicit def setEncoder[A: ScynamoEncoder]: ScynamoEncoder[Set[A]] =
+    listEncoder[A].contramap(_.toList)
 
-  implicit def optionEncoder[A: ScynamoEncoder]: ScynamoEncoder[Option[A]] = {
-    case Some(value) => ScynamoEncoder[A].encode(value)
-    case None        => Right(AttributeValue.builder().nul(true).build())
-  }
+  implicit def optionEncoder[A](implicit element: ScynamoEncoder[A]): ScynamoEncoder[Option[A]] =
+    ScynamoEncoder.instance {
+      case Some(value) => element.encode(value)
+      case None        => rightNul
+    }
 
-  implicit def someEncoder[A: ScynamoEncoder]: ScynamoEncoder[Some[A]] = x => ScynamoEncoder[A].encode(x.get)
+  implicit def someEncoder[A](implicit element: ScynamoEncoder[A]): ScynamoEncoder[Some[A]] =
+    ScynamoEncoder.instance(some => element.encode(some.get))
 
-  implicit val finiteDurationEncoder: ScynamoEncoder[FiniteDuration] = longEncoder.contramap(_.toNanos)
+  implicit val finiteDurationEncoder: ScynamoEncoder[FiniteDuration] =
+    numberStringEncoder.contramap(_.toNanos.toString)
 
-  implicit val durationEncoder: ScynamoEncoder[Duration] = longEncoder.contramap(_.toNanos)
+  implicit val durationEncoder: ScynamoEncoder[Duration] =
+    numberStringEncoder.contramap(_.toNanos.toString)
 
-  implicit def mapEncoder[A, B](implicit keyEncoder: ScynamoKeyEncoder[A], valueEncoder: ScynamoEncoder[B]): ScynamoEncoder[Map[A, B]] =
-    value => {
-      value.toVector
-        .parTraverse { case (k, v) =>
-          (keyEncoder.encode(k), valueEncoder.encode(v)).parMapN(_ -> _).leftMap(_.map(_.push(MapKey(k))))
-        }
-        .map {
-          _.foldLeft(new java.util.HashMap[String, AttributeValue]()) { case (acc, (k, v)) =>
-            acc.put(k, v)
-            acc
-          }
+  implicit def mapEncoder[A, B](implicit key: ScynamoKeyEncoder[A], value: ScynamoEncoder[B]): ScynamoEncoder[Map[A, B]] =
+    ScynamoEncoder.instance { kvs =>
+      var allErrors  = Chain.empty[ScynamoEncodeError]
+      val attrValues = new java.util.HashMap[String, AttributeValue](kvs.size)
+      kvs.foreachEntry { (k, v) =>
+        (key.encode(k), value.encode(v)).parTupled match {
```

Comment:
If we go for efficiency, we should probably remove the `parTupled` here as well and just pattern match all the cases

Reply:
Probably 👍